sparse/compressed/compressed.py

```diff
@@ -0,0 +1,329 @@
+            compressed_axes = (np.argmin(self.shape),)
+        elif len(compressed_axes) >= len(shape):
+            raise ValueError('cannot compress all axes')
+
+        compressed_axes = tuple(np.unique(compressed_axes))
+        axis_order = list(compressed_axes)
+        axisptr = len(compressed_axes) # array location where the uncompressed dimensions start
+        axis_order.extend(np.setdiff1d(np.arange(len(shape)),compressed_axes))
+        reordered_shape = np.array(shape)[axis_order]
+        row_size = np.prod(reordered_shape[:axisptr])
+        col_size = np.prod(reordered_shape[axisptr:])
+        compressed_shape = (row_size,col_size)
+
+        if isinstance(arg,tuple):
+            self.data,self.indices,self.indptr = arg
+            self.shape = shape
+            self.compressed_shape = compressed_shape
+            self.compressed_axes = compressed_axes
+            self.axis_order = axis_order
+            self.axisptr = axisptr
+            self.reordered_shape = reordered_shape
+            self.fill_value = fill_value
+            self.dtype = self.data.dtype
+
+    @classmethod
+    def from_numpy(cls,x,compressed_axes=None,fill_value=0):
+        coo = COO(x,fill_value=fill_value)
+        return cls.from_coo(coo,compressed_axes)
+
+
+    @classmethod
+    def from_coo(cls,x,compressed_axes=None):
+        arg, shape, compressed_shape,compressed_axes,axis_order,reordered_shape,axisptr,fill_value = _from_coo(x,compressed_axes)
+        return cls(arg,shape=shape, compressed_axes=compressed_axes,fill_value=fill_value)
+
+    @classmethod
+    def from_scipy_sparse(cls,x):
+        if x.format == 'csc':
+            return cls((x.data,x.indices,x.indptr),shape=x.shape,compressed_axes=(1,))
+        else:
+            x = x.asformat('csr')
+            return cls((x.data,x.indices,x.indptr),shape=x.shape,compressed_axes=(0,))
+        
+
+    @classmethod
+    def from_iter(cls,x,shape=None,compressed_axes=None,fill_value=None):
+        return cls.from_coo(COO.from_iter(x,shape,fill_value),compressed_axes=compressed_axes)
+
+    @property
+    def nnz(self):
+        return self.data.shape[0]
+
+    @property
+    def nbytes(self):
+        return self.data.nbytes + self.indices.nbytes + self.indptr.nbytes
+    
+    @property
+    def density(self):
+        return self.nnz / reduce(mul,self.shape)
+
+    @property
+    def ndim(self):
+        return len(self.shape)
+
+    def __str__(self):
+        return '<CSD: shape={}, dtype={}, nnz={}, fill_value={}, compressed_axes={}>'.format(self.shape,
+            self.dtype,self.nnz,self.fill_value,self.compressed_axes)
+
+    __repr__ = __str__      
+
+
+    __getitem__ = getitem
+
+
+    def change_compressed_axes(self,new_compressed_axes):
+        """
+        changes the compressed axes in-place.
+        right now the space complexity feels a little high
+        """
+        new_compressed_axes = tuple(np.unique(new_compressed_axes))

```

Comment:
Don't make them unique, there's absolutely no need. 😄 

Instead _check_ for uniqueness, that should be the requirement.

Same in the constructor.